File: python/open_guardrail/guards/url_phishing.py
"""Detect phishing URLs including homograph attacks."""
from __future__ import annotations

import re
import time
from urllib.parse import urlparse

from open_guardrail.core import GuardResult

_CYRILLIC_LATIN = {
    "\u0430": "a", "\u0435": "e",
    "\u043e": "o", "\u0440": "p",
    "\u0441": "c", "\u0443": "y",
    "\u0445": "x", "\u0456": "i",
}

_URL_RE = re.compile(r"https?://[^\s\"'<>]+", re.I)

_SUSPICIOUS_SUBS = {
    "login", "signin", "secure", "account",
    "verify", "update", "confirm", "bank",
}


def _has_homograph(url: str) -> bool:
    return any(c in _CYRILLIC_LATIN for c in url)


def _suspicious_subdomain(url: str) -> bool:
    try:
        host = urlparse(url).hostname or ""
        parts = host.split(".")
        if len(parts) <= 2:
            return False
        return any(
            p in _SUSPICIOUS_SUBS for p in parts[:-2]
        )
    except Exception:
        return False


def _is_ip_based(url: str) -> bool:
    try:
        host = urlparse(url).hostname or ""
        return bool(
            re.fullmatch(r"\d{1,3}(\.\d{1,3}){3}", host)
        )
    except Exception:
        return False


def _excessive_depth(url: str) -> bool:
    try:
        path = urlparse(url).path
        return len(
            [s for s in path.split("/") if s]
        ) > 8
    except Exception:
        return False


_DATA_URI_RE = re.compile(
    r"data:[^;]+;base64,", re.I
)


class _UrlPhishing:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "url-phishing"
        self.action = action

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        issues: list[str] = []

        if _DATA_URI_RE.search(text):
            issues.append("data_uri")

        for m in _URL_RE.finditer(text):
            url = m.group(0)
            if _has_homograph(url):
                issues.append("homograph_attack")
            if _suspicious_subdomain(url):
                issues.append("suspicious_subdomain")
            if _is_ip_based(url):
                issues.append("ip_based_url")
            if _excessive_depth(url):
                issues.append("excessive_depth")

        triggered = len(issues) > 0
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="url-phishing",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=(
                min(len(issues) / 3, 1.0)
                if triggered
                else 0.0
            ),
            message=(
                "Phishing URL detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"issues": issues} if triggered else None
            ),
        )


def url_phishing(
    *, action: str = "block"
) -> _UrlPhishing:
    return _UrlPhishing(action=action)

File: packages/guards/src/url-phishing.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface UrlPhishingOptions {
  action: 'block' | 'warn';
}

const CYRILLIC_LATIN_MAP: Record<string, string> = {
  '\u0430': 'a', '\u0435': 'e', '\u043e': 'o', '\u0440': 'p',
  '\u0441': 'c', '\u0443': 'y', '\u0445': 'x', '\u0456': 'i',
};

const URL_RE = /https?:\/\/[^\s"'<>]+/gi;

function hasHomograph(url: string): boolean {
  for (const char of url) {
    if (CYRILLIC_LATIN_MAP[char]) return true;
  }
  return false;
}

function hasSuspiciousSubdomain(url: string): boolean {
  try {
    const hostname = new URL(url).hostname;
    const parts = hostname.split('.');
    if (parts.length <= 2) return false;
    const suspicious = ['login', 'signin', 'secure', 'account', 'verify', 'update', 'confirm', 'bank'];
    return parts.slice(0, -2).some(p => suspicious.includes(p));
  } catch {
    return false;
  }
}

function isIpBased(url: string): boolean {
  try {
    const hostname = new URL(url).hostname;
    return /^\d{1,3}(\.\d{1,3}){3}$/.test(hostname);
  } catch {
    return false;
  }
}

function hasExcessiveDepth(url: string): boolean {
  try {
    const path = new URL(url).pathname;
    return path.split('/').filter(Boolean).length > 8;
  } catch {
    return false;
  }
}

function isDataUri(text: string): boolean {
  return /data:[^;]+;base64,/i.test(text);
}

export function urlPhishing(options: UrlPhishingOptions): Guard {
  return {
    name: 'url-phishing',
    version: '0.1.0',
    description: 'Detects phishing URLs including homograph attacks and suspicious patterns',
    category: 'security',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const issues: string[] = [];

      if (isDataUri(text)) {
        issues.push('data_uri');
      }

      let match: RegExpExecArray | null;
      const re = new RegExp(URL_RE.source, URL_RE.flags);
      while ((match = re.exec(text)) !== null) {
        const url = match[0];
        if (hasHomograph(url)) issues.push('homograph_attack');
        if (hasSuspiciousSubdomain(url)) issues.push('suspicious_subdomain');
        if (isIpBased(url)) issues.push('ip_based_url');
        if (hasExcessiveDepth(url)) issues.push('excessive_depth');
      }

      const triggered = issues.length > 0;
      return {
        guardName: 'url-phishing',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score: triggered ? Math.min(issues.length / 3, 1.0) : 0,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { issues } : undefined,
      };
    },
  };
}
